core/src/filesystem/media/analyze_media_job.rs

```diff
@@ -0,0 +1,184 @@
+use std::collections::VecDeque;
+
+use serde::{Deserialize, Serialize};
+use specta::Type;
+
+use crate::{
+	filesystem::media::process::get_page_count,
+	job::{
+		error::JobError, JobExecuteLog, JobExt, JobOutputExt, JobTaskOutput, WorkerCtx,
+		WorkingState, WrappedJob,
+	},
+	prisma::{media, series},
+};
+
+#[derive(Clone)]
+pub enum AnalyzeMediaJobVariant {
+	AnalyzeSingleItem(String),
+	AnalyzeLibrary(String),
+	AnalyzeSeries(String),
+	AnalyzeMediaGroup(Vec<String>),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum AnalyzeMediaTask {
+	AnalyzeImage(String),
+}
+
+#[derive(Clone, Serialize, Deserialize, Default, Debug, Type)]
+pub struct AnalyzeMediaOutput {
+	/// The number of images analyzed
+	images_analyzed: u64,
+}
+
+impl JobOutputExt for AnalyzeMediaOutput {
+	fn update(&mut self, updated: Self) {
+		self.images_analyzed += updated.images_analyzed;
+	}
+}
+
+/// A job that analyzes a media item and updates the database
+/// with information from the analysis.
+#[derive(Clone)]
+pub struct AnalyzeMediaJob {
+	pub variant: AnalyzeMediaJobVariant,
+}
+
+impl AnalyzeMediaJob {
+	/// Create a new [AnalyzeMediaJob] for the media specified by `id`.
+	pub fn new(variant: AnalyzeMediaJobVariant) -> Box<WrappedJob<AnalyzeMediaJob>> {
+		WrappedJob::new(Self { variant })
+	}
+}
+
+#[async_trait::async_trait]
+impl JobExt for AnalyzeMediaJob {
+	const NAME: &'static str = "analyze_media";
+
+	type Output = AnalyzeMediaOutput;
+	type Task = AnalyzeMediaTask;
+
+	fn description(&self) -> Option<String> {
+		match &self.variant {
+			AnalyzeMediaJobVariant::AnalyzeSingleItem(id) => {
+				Some(format!("Analyze media item with id: {}", id))
+			},
+			AnalyzeMediaJobVariant::AnalyzeLibrary(id) => {
+				Some(format!("Analyze library with id: {}", id))
+			},
+			AnalyzeMediaJobVariant::AnalyzeSeries(id) => {
+				Some(format!("Analyze series with id: {}", id))
+			},
+			AnalyzeMediaJobVariant::AnalyzeMediaGroup(ids) => {
+				Some(format!("Analyze media group with ids: {:?}", ids))
+			},
+		}
+	}
+
+	async fn init(
+		&mut self,
+		ctx: &WorkerCtx,
+	) -> Result<WorkingState<Self::Output, Self::Task>, JobError> {
+		tracing::warn!("In AnalyzeMediaJob init()");
+		let output = Self::Output::default();
+
+		// We match over the job variant to build a list of tasks to process
+		let mut tasks = VecDeque::new();
+		match &self.variant {
+			// Single item is easy
+			AnalyzeMediaJobVariant::AnalyzeSingleItem(id) => {
+				tasks.push_front(AnalyzeMediaTask::AnalyzeImage(id.clone()))
+			},
+			// For libraries we need a list of ids
+			AnalyzeMediaJobVariant::AnalyzeLibrary(id) => {
+				let library_media = ctx
+					.db
+					.media()
+					.find_many(vec![media::series::is(vec![series::library_id::equals(
+						Some(id.clone()),
+					)])])
+					.exec()
+					.await
+					.map_err(|e| JobError::InitFailed(e.to_string()))?;
```

Comment:
I think I follow the problem you're describing here - we load a lot of data here when all we need is the ids, but I'm not sure of the right way to implement the solution. You linked this example:

```rust
let books = tx
	.media()
	.find_many(params)
	.select(media_only_series_id::select())
	.exec()
	.await?;
```

Does `media_only_series_id::select()` cause only the `series_id` to be loaded? If that's right, how could I instead only load the media id? This is really just my lack of familiarity with prisma.

Reply:
> Does `media_only_series_id::select()` cause only the series_id to be loaded?

Yep! It does exactly that. Here are the [relevant docs](https://prisma.brendonovich.dev/reading-data/select-include), but offhand it would be something like this if we needed to capture and export it:

```rust
media::select!(media_ids_only {
   id
});
```

I generally place a `prisma_macros.rs` file adjacent to the entity in the `db` module of `stump_core`. This is because of limitations with how the path resolution works while using the select/include macros, which prevents using it outside the crate (e.g. `stump_server`). It's a bit magical, but the docs will explain better than I could probably.

In this instance, we probably don't even need a named select though since we are using it in the core and not anywhere else:

```rust
let library_media_ids = ctx
	.db
	.media()
	.find_many(vec![media::series::is(vec![series::library_id::equals(
		Some(id.clone()),
	)])])
	.select(media::select!({
		id
	}))
	.exec()
	.await
	.map_err(|e| JobError::InitFailed(e.to_string()))?;
```

Reply:
That works great, easy fix - thanks for explaining.